File: cpp/853_car_fleet.cpp
#include <utility>
#include <vector>

using namespace std;

class Solution {
 public:
  int carFleet(int target, vector<int>& position, vector<int>& speed) {
    int n = position.size();
    vector<pair<int, float>> arrive_time(n);
    for (int i = 0; i < n; ++i) {
      float time = (target - position[i]) / float(speed[i]);
      arrive_time[i] = {position[i], time};
    }

    sort(arrive_time.begin(), arrive_time.end());

    int result = 0;
    float max_time = 0;
    for (int i = n - 1; i > -1; --i) {
      float time = arrive_time[i].second;
      if (time > max_time) {
        max_time = time;
        result++;
      }
    }
    return result;
  }
};

File: rust/853_car_fleet.rs
use std::cmp::Ordering;

impl Solution {
    pub fn car_fleet(target: i32, position: Vec<i32>, speed: Vec<i32>) -> i32 {
        let mut arrive_time: Vec<(i32, f32)> = position
            .into_iter()
            .zip(speed)
            .map(|(pos, speed)| (pos, (target - pos) as f32 / speed as f32))
            .collect();

        arrive_time.sort_unstable_by_key(|x| x.0);

        let mut result = 0;
        let mut max_time = 0f32;

        while let Some((_, time)) = arrive_time.pop() {
            if time > max_time {
                max_time = time;
                result += 1;
            }
        }

        result
    }
}
